noobaa_sa/s3_client.py

```diff
@@ -0,0 +1,293 @@
+            prefix (str): The S3 prefix to download from (acts like a directory).
+        """
+
+        transfer_config = TransferConfig(use_threads=True)
+
+        log.info(f"Downloading s3:///{bucket_name}/{prefix} to {local_dir} via boto3")
+        # List objects within the specified prefix
+        for obj in self.list_objects(bucket_name, prefix, use_v2=True):
+            # Construct the full local path
+            relative_path = os.path.relpath(obj, prefix)
+            local_file_path = os.path.join(local_dir, relative_path)
+
+            # Ensure local directory structure mirrors S3
+            local_file_dir = os.path.dirname(local_file_path)
+            if not os.path.exists(local_file_dir):
+                os.makedirs(local_file_dir)
+
+            print(f"Downloading {obj} to {local_file_path}")
+            self._boto3_client.download_file(
+                bucket_name, obj, local_file_path, Config=transfer_config
+            )
+
+    def delete_object(self, bucket_name, object_key):
+        """
+        Delete an object from an S3 bucket using boto3
+
+        """
+        log.info(f"Deleting object {object_key} from bucket {bucket_name} via boto3")
+        self._boto3_client.delete_object(Bucket=bucket_name, Key=object_key)
+
+    def put_random_objects(
+        self, bucket_name, amount=10, obj_size="1M", prefix="", files_dir=""
+    ):
+        """
+        Write random objects to an S3 bucket
+
+        Args:
+            bucket_name (str): The name of the bucket to write to
+            amount (int): The number of objects to write
+            obj_size (str): The size of each object
+            prefix (str): A prefix where the objects will be written in the bucket
+            files_dir (str): A directory where the objects will be written locally.
+                             If not specified, a temporary directory will be used.
+
+        Returns:
+            list: A list of the names of the objects written to the bucket
+
+        """
+        written_objs = []
+
+        # Determine the directory to use for file storage
+        actual_files_dir = files_dir if files_dir else tempfile.mkdtemp()
+
+        # Generate random files in the specified directory
+        written_objs += generate_random_files(actual_files_dir, amount, obj_size)
+
+        # Ensure the prefix ends with a slash if it is not empty
+        if prefix and not prefix.endswith("/"):
+            prefix += "/"
+
+        self.upload_directory(actual_files_dir, bucket_name, prefix)
+
+        return written_objs
+
+    def delete_all_objects_in_bucket(self, bucket_name):
+        """
+        Deletes all objects in the specified S3 bucket.
+
+        Args:
+            bucket_name (str): The name of the S3 bucket.
+        """
+        # TODO: Support buckets with versioning enabled
+        log.info(f"Deleting all objects in bucket {bucket_name} via boto3")
+        self._boto3_resource.Bucket(bucket_name).objects.all().delete()
```

Comment:
We need to add sleep with condition here since we are immediately calling next operation without waiting for clean up. This might end up getting `BucketNotEmpty` error if bucket is having more objects in next step.

Reply:
This is also relevant for many other operations, and not just in S3Client. We should address this in a more general way in the future, but for now we can add waiting and validation in the tests that use these operations.

I've created the following issue to track this: https://github.com/red-hat-storage/noobaa-sa-ci/issues/12